{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env bash\n",
    "#SBATCH -A SNIC2019-4-10\n",
    "#SBATCH -p hebbe\n",
    "#SBATCH -J aligner\n",
    "#SBATCH -n 2\n",
    "#SBATCH -t 0-08:00:00\n",
    "\n",
    "module load Anaconda3/5.3.0;\n",
    "source activate /c3se/NOBACKUP/users/mariaji/envs/bwa;\n",
    "module load iccifort/2019.1.144-GCC-8.2.0-2.31.1;\n",
    "module load SAMtools/1.9;\n",
    "module load GATK/3.7-Java-1.8.0_112;\n",
    "module load picard/2.18.25-Java-1.8;\n",
    "\n",
    "# bwa index -a bwtsw ${REF};\n",
    "# samtools faidx ${REF}\n",
    "\n",
    "# java -jar $EBROOTPICARD/picard.jar CreateSequenceDictionary \\\n",
    "# REFERENCE=${REF} \\\n",
    "# OUTPUT=/c3se/NOBACKUP/users/mariaji/palms_exons_final.dict;\n",
    "\n",
    "# mkdir rawbams;\n",
    "# mkdir indelvcf;\n",
    "# mkdir realnbam;\n",
    "# mkdir rawvcf;\n",
    "# mkdir recal1;\n",
    "# mkdir recal2;\n",
    "# mkdir variants;\n",
    "\n",
    "#######################################################################\n",
    "# 1. map nodup reads to loci reference abd prepare bam\n",
    "#######################################################################\n",
    "\n",
    "# modify depending on data\n",
    "# running from outside folder with index folders\n",
    "# reads are trimmed and PCR duplicates masked with Picard\n",
    "SAMPLE=$(ls Index${SLURM_ARRAY_TASK_ID}/*_nodup_clean-READ1.fastq | sed -r 's/_nodup_clean-READ1.fastq//g');\n",
    "GROUP=$(echo ${SAMPLE} | sed -r 's/Index[0-9]+\\///g')\n",
    "REF=$(path/to/ref.fasta)\n",
    "# echo ${SAMPLE} >> text.txt\n",
    "# echo ${GROUP} >> text.txt\n",
    "\n",
    "bwa mem -t 4 ${REF} \\\n",
    "    ${SAMPLE}_nodup_clean-READ1.fastq \\\n",
    "    ${SAMPLE}_nodup_clean-READ2.fastq \\\n",
    "    -o rawbams/${SAMPLE}.sam $2> rawbams/${SAMPLE}_step1.out;\n",
    "\n",
    "java -jar $EBROOTPICARD/picard.jar SamFormatConverter \\\n",
    "    I=rawbams/${SAMPLE}.sam \\\n",
    "    O=rawbams/${SAMPLE}.bam;\n",
    "\n",
    "java -jar $EBROOTPICARD/picard.jar SortSam \\\n",
    "      I=rawbams/${SAMPLE}.bam \\\n",
    "      O=rawbams/${SAMPLE}.sorted.bam \\\n",
    "      SORT_ORDER=coordinate\n",
    "\n",
    "java -jar $EBROOTPICARD/picard.jar BuildBamIndex \\\n",
    "      I=rawbams/${SAMPLE}.sorted.bam;\n",
    "\n",
    "# when automatising this make sure that the ${SAMPLE} variable excludes folder paths\n",
    "java -jar $EBROOTPICARD/picard.jar AddOrReplaceReadGroups \\\n",
    "       I=rawbams/${SAMPLE}.sorted.bam \\\n",
    "       O=rawbams/${SAMPLE}.sorted.wgroups.bam \\\n",
    "       RGID=4 \\\n",
    "       RGLB=lib1 \\\n",
    "       RGPL=illumina \\\n",
    "       RGPU=unit1 \\\n",
    "       RGSM=${GROUP} $2> rawbams/${SAMPLE}_groupfix.out;\n",
    "\n",
    "java -jar $EBROOTPICARD/picard.jar BuildBamIndex \\\n",
    "      I=rawbams/${SAMPLE}.sorted.wgroups.bam;\n",
    "\n",
    "rm rawbams/${SAMPLE}.sorted.bam;\n",
    "rm rawbams/${SAMPLE}.sorted.bai;\n",
    "rm rawbams/${SAMPLE}.sam;\n",
    "rm rawbams/${SAMPLE}.bam;\n",
    "\n",
    "echo '**** Index built and bam-sam-bam switch performed****';\n",
    "\n",
    "#######################################################################\n",
    "# 2. call first VCF\n",
    "#######################################################################\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "   -T UnifiedGenotyper \\\n",
    "   -R ${REF} \\\n",
    "   -I rawbams/${SAMPLE}.sorted.wgroups.bam \\\n",
    "   -o rawvcf/${SAMPLE}.raw.vcf $2> rawvcf/${SAMPLE}_step2.out;\n",
    "\n",
    "echo '****raw variants called****';\n",
    "\n",
    "######################################################################\n",
    "# 3. re-align indels\n",
    "######################################################################\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "   -T SelectVariants \\\n",
    "   -R ${REF} \\\n",
    "   -V rawvcf/${SAMPLE}.raw.vcf \\\n",
    "   -o indelvcf/${SAMPLE}_indelraw.vcf \\\n",
    "   -selectType INDEL \\\n",
    "   --minIndelSize 2 \\\n",
    "   --maxIndelSize 50 $2> indelvcf/${SAMPLE}_step3_callindel.out;\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "   -T RealignerTargetCreator \\\n",
    "   -R ${REF} \\\n",
    "   -I rawbams/${SAMPLE}.sorted.wgroups.bam \\\n",
    "   --known indelvcf/${SAMPLE}_indelraw.vcf \\\n",
    "   -o indelvcf/${SAMPLE}.indel.intervals $2> indelvcf/${SAMPLE}_step3_indelinterv.out;\n",
    "\n",
    "# targetIntervals output from RealignerTargetCreator\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "   -T IndelRealigner \\\n",
    "   -R ${REF} \\\n",
    "   -I rawbams/${SAMPLE}.sorted.wgroups.bam \\\n",
    "   -known indelvcf/${SAMPLE}_indelraw.vcf \\\n",
    "   -targetIntervals indelvcf/${SAMPLE}.indel.intervals \\\n",
    "   -o realnbam/${SAMPLE}.realn.bam $2> indelvcf/${SAMPLE}_step3_realn.out;\n",
    "# need to deal with \"Not attempting realignment in interval ... because there are too many reads.\"\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "   -T UnifiedGenotyper \\\n",
    "   -R ${REF} \\\n",
    "   -I realnbam/${SAMPLE}.realn.bam \\\n",
    "   -o realnbam/${SAMPLE}.realn.vcf $2> realnbam/${SAMPLE}_step3_ralnvcf.out;\n",
    "\n",
    "rm rawbams/${SAMPLE}.sorted.wgroups.bam;\n",
    "rm rawbams/${SAMPLE}.sorted.wgroups.bai;\n",
    "rm indelvcf/${SAMPLE}_indelraw.vcf;\n",
    "rm indelvcf/${SAMPLE}.indel.intervals;\n",
    "\n",
    "#######################################################################\n",
    "# 4. recalibration round 1 - ideally run two minimum\n",
    "#######################################################################\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "    -T BaseRecalibrator \\\n",
    "    -R ${REF} \\\n",
    "    -I realnbam/${SAMPLE}.realn.bam \\\n",
    "    -knownSites realnbam/${SAMPLE}.realn.vcf \\\n",
    "    -o recal1/${SAMPLE}_recal_data.1.table $2> recal1/${SAMPLE}_step4_recal1.out;\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "    -T BaseRecalibrator \\\n",
    "    -R ${REF} \\\n",
    "    -I realnbam/${SAMPLE}.realn.bam \\\n",
    "    -knownSites realnbam/${SAMPLE}.realn.vcf \\\n",
    "    -BQSR recal1/${SAMPLE}_recal_data.1.table \\\n",
    "    -o recal1/${SAMPLE}_post_recal_data.1.table $2> recal1/${SAMPLE}_step4_recal1post.out; \n",
    "    \n",
    "#######################################################################\n",
    "# 5. filter with recalibration and call gVCF\n",
    "#######################################################################\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "   -T PrintReads \\\n",
    "   -R ${REF} \\\n",
    "   -I realnbam/${SAMPLE}.realn.bam \\\n",
    "   -BQSR recal1/${SAMPLE}_post_recal_data.1.table \\\n",
    "   -o recal1/${SAMPLE}.recab.bam $2> recal1/${SAMPLE}_step4_recab.out;\n",
    "\n",
    "java -jar $EBROOTGATK/GenomeAnalysisTK.jar \\\n",
    "   -T UnifiedGenotyper \\\n",
    "   -R ${REF} \\\n",
    "   -I recal1/${SAMPLE}.recab.bam \\\n",
    "   -o variants/${SAMPLE}.nodup.recab.realn.vcf \\\n",
    "   --output_mode EMIT_ALL_SITES $2> variants/${SAMPLE}_varcal.out;\n",
    "\n",
    "rm realnbam/${SAMPLE}.realn.bam\n",
    "\n",
    "# rm realnbam/${SAMPLE}_realn.bam;\n",
    "touch \"${SAMPLE}\"_done.out;"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python36",
   "language": "python",
   "name": "python36"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
